src/lib/u2/browser.js

```diff
@@ -0,0 +1,550 @@
+      }
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.u2',
+  name: 'DAOUpdateController',
+  extends: 'foam.u2.View',
+  requires: [
+    'foam.u2.DetailView'
+  ],
+
+  imports: [
+    'document',
+    'stack'
+  ],
+
+  exports: [
+    'myControllerMode as controllerMode'
+  ],
+
+  properties: [
+    {
+      // The original, golden data. innerData is a clone of this.
+      name: 'data',
+      postSet: function(old, nu) {
+        // Only overwrite innerData if it's clean relative to old.
+        // That prevents overwriting any changes the user might have made.
+        if ( old && old.equals(this.innerData) ) {
+          this.innerData = nu.clone();
+        }
+      }
+    },
+    {
+      name: 'innerData',
+      factory: function() {
+        return this.data.clone();
+      },
+      postSet: function(old, nu) {
+        if ( old === nu ) return;
+        if ( this.dataSub_ ) this.dataSub_.destroy();
+        this.dataSub_ = nu.propertyChange.sub(this.onDataChange);
+        this.onDataChange();
+      }
+    },
+    'dataSub_',
+    {
+      name: 'of',
+      expression: function(data) {
+        return data.cls_;
+      }
+    },
+    {
+      name: 'dao',
+      expression: function(of) {
+        return this.__context__[foam.String.daoize(of.name)];
+      }
+    },
+    {
+      name: 'title',
+      expression: function(of) {
+        return 'Edit ' + of.name;
+      }
+    },
+    {
+      class: 'Boolean',
+      name: 'dirty_'
+    },
+    [ 'myControllerMode', 'update' ]
+  ],
+
+  actions: [
+    {
+      name: 'back',
+      isAvailable: function(dirty_) { return ! dirty_; },
+      code: function(X) {
+        X.stack.popMe(this);
+      }
+    },
+    {
+      name: 'delete',
+      isAvailable: function(dirty_) { return ! dirty_; },
+      code: function(X) {
+        this.dao.remove(this.data);
+        this.back();
+      }
+    },
+    {
+      name: 'cancel',
+      isAvailable: function(dirty_) { return dirty_; },
+      code: function(X) {
+        this.clearFocus_();
+        this.innerData = this.data.clone();
+      }
+    },
+    {
+      name: 'submit',
+      isAvailable: function(dirty_) { return dirty_; },
+      code: function(X) {
+        // TODO(braden): Handle long-running submits here. Some UI affordance
+        // for the fact that work is being done here.
+        this.clearFocus_();
+        this.dao.put(this.innerData).then(function(nu) {
+          this.data = nu;
+          this.innerData = nu.clone();
+        }.bind(this));
+      }
+    }
+  ],
+
+  methods: [
+    function initE() {
+      var dv = this.DetailView.create({ of: this.of });
+      dv.data$ = this.innerData$;
+      var X = this.__subContext__.createSubContext({ data: this });
+      this.add(this.BACK.toE(X));
+      this.add(this.CANCEL.toE(X));
+      this.add(this.DELETE.toE(X));
+      this.add(this.SUBMIT.toE(X));
+      this.add(dv);
+      this.enableCls('red-border', this.slot('dirty_'));
```

Comment:
Reminds me that we should rename cssClass to just cls() since we are inconsistent with myCls() and enableCls().